File: CurrencyFavorites.cs
// CurrencyFavorites.cs
using System;
using System.Collections.Generic;
using System.IO;
using System.Linq;
using System.Net.Http;
using System.Text.Json;
using System.Text.Json.Serialization;
using System.Threading.Tasks;

class Pair
{
    [JsonPropertyName("base")] public string Base { get; set; }
    [JsonPropertyName("target")] public string Target { get; set; }
}

class Favorites
{
    [JsonPropertyName("pairs")] public List<Pair> Pairs { get; set; } = new List<Pair>();
}

class CurrencyFavorites
{
    private static readonly string FavoritesFile = "favorites.json";
    private static readonly string ApiUrl = "https://api.frankfurter.app/latest?from=";
    private static readonly HttpClient client = new HttpClient();
    private static readonly JsonSerializerOptions options = new JsonSerializerOptions { WriteIndented = true };

    static async Task<List<Pair>> LoadFavoritesAsync()
    {
        if (File.Exists(FavoritesFile))
        {
            string json = await File.ReadAllTextAsync(FavoritesFile);
            var fav = JsonSerializer.Deserialize<Favorites>(json);
            return fav?.Pairs ?? new List<Pair>();
        }
        return new List<Pair> { new Pair { Base = "USD", Target = "EUR" }, new Pair { Base = "USD", Target = "JPY" } };
    }

    static async Task SaveFavoritesAsync(List<Pair> pairs)
    {
        var fav = new Favorites { Pairs = pairs };
        string json = JsonSerializer.Serialize(fav, options);
        await File.WriteAllTextAsync(FavoritesFile, json);
    }

    static async Task<Dictionary<string, double>> GetRatesAsync(string baseCurrency)
    {
        string url = ApiUrl + baseCurrency;
        var response = await client.GetAsync(url);
        response.EnsureSuccessStatusCode();
        string json = await response.Content.ReadAsStringAsync();
        using var doc = JsonDocument.Parse(json);
        var ratesElement = doc.RootElement.GetProperty("rates");
        var dict = new Dictionary<string, double>();
        foreach (var prop in ratesElement.EnumerateObject())
        {
            dict[prop.Name] = prop.Value.GetDouble();
        }
        return dict;
    }

    static async Task ListRatesAsync(List<Pair> pairs)
    {
        if (pairs.Count == 0)
        {
            Console.WriteLine("No favorite pairs.");
            return;
        }
        var grouped = pairs.GroupBy(p => p.Base);
        Console.WriteLine("\n💱 Favorite Currency Pairs\n");
        foreach (var group in grouped)
        {
            string baseCurrency = group.Key;
            try
            {
                var rates = await GetRatesAsync(baseCurrency);
                foreach (var p in group)
                {
                    if (rates.TryGetValue(p.Target, out double rate))
                    {
                        Console.WriteLine($"{p.Base}/{p.Target}: {rate:F4}");
                    }
                    else
                    {
                        Console.WriteLine($"{p.Base}/{p.Target}: N/A");
                    }
                }
            }
            catch (Exception e)
            {
                Console.WriteLine($"Error fetching rates for {baseCurrency}: {e.Message}");
            }
        }
        Console.WriteLine($"\nUpdated: {DateTime.Now:yyyy-MM-dd HH:mm:ss}");
    }

    static async Task Main(string[] args)
    {
        if (args.Length == 0)
        {
            var pairs = await LoadFavoritesAsync();
            await ListRatesAsync(pairs);
            return;
        }
        string cmd = args[0].ToLower();
        switch (cmd)
        {
            case "list":
                pairs = await LoadFavoritesAsync();
                await ListRatesAsync(pairs);
                break;
            case "add":
                if (args.Length != 3)
                {
                    Console.WriteLine("Usage: add BASE TARGET");
                    return;
                }
                string baseCur = args[1].ToUpper();
                string targetCur = args[2].ToUpper();
                pairs = await LoadFavoritesAsync();
                if (!pairs.Any(p => p.Base == baseCur && p.Target == targetCur))
                {
                    pairs.Add(new Pair { Base = baseCur, Target = targetCur });
                    await SaveFavoritesAsync(pairs);
                    Console.WriteLine($"✅ Added {baseCur}/{targetCur}");
                }
                else
                {
                    Console.WriteLine($"Pair {baseCur}/{targetCur} already in favorites.");
                }
                break;
            case "remove":
                if (args.Length != 3)
                {
                    Console.WriteLine("Usage: remove BASE TARGET");
                    return;
                }
                baseCur = args[1].ToUpper();
                targetCur = args[2].ToUpper();
                pairs = await LoadFavoritesAsync();
                bool removed = pairs.RemoveAll(p => p.Base == baseCur && p.Target == targetCur) > 0;
                if (removed)
                {
                    await SaveFavoritesAsync(pairs);
                    Console.WriteLine($"✅ Removed {baseCur}/{targetCur}");
                }
                else
                {
                    Console.WriteLine($"Pair {baseCur}/{targetCur} not found.");
                }
                break;
            case "watch":
                int interval = 60;
                for (int i = 1; i < args.Length; i++)
                {
                    if (args[i] == "--interval" && i + 1 < args.Length)
                    {
                        interval = int.Parse(args[i + 1]);
                        i++;
                    }
                }
                Console.WriteLine($"Watching every {interval}s. Press Ctrl+C to stop.");
                var watchPairs = await LoadFavoritesAsync();
                while (true)
                {
                    await ListRatesAsync(watchPairs);
                    await Task.Delay(interval * 1000);
                }
            default:
                Console.WriteLine("Usage: CurrencyFavorites [list|add BASE TARGET|remove BASE TARGET|watch [--interval N]]");
                break;
        }
    }
}

File: currency_favorites.cpp
// currency_favorites.cpp
#include <iostream>
#include <fstream>
#include <string>
#include <vector>
#include <map>
#include <thread>
#include <chrono>
#include <curl/curl.h>
#include <nlohmann/json.hpp>

using namespace std;
using json = nlohmann::json;

const string FAVORITES_FILE = "favorites.json";
const string API_URL = "https://api.frankfurter.app/latest?from=";

struct Pair {
    string base;
    string target;
};

size_t WriteCallback(void *contents, size_t size, size_t nmemb, void *userp) {
    ((string*)userp)->append((char*)contents, size * nmemb);
    return size * nmemb;
}

string fetchUrl(const string& url) {
    CURL *curl = curl_easy_init();
    string response;
    if (curl) {
        curl_easy_setopt(curl, CURLOPT_URL, url.c_str());
        curl_easy_setopt(curl, CURLOPT_WRITEFUNCTION, WriteCallback);
        curl_easy_setopt(curl, CURLOPT_WRITEDATA, &response);
        curl_easy_setopt(curl, CURLOPT_TIMEOUT, 10L);
        CURLcode res = curl_easy_perform(curl);
        curl_easy_cleanup(curl);
        if (res != CURLE_OK) return "";
    }
    return response;
}

vector<Pair> loadFavorites() {
    vector<Pair> pairs;
    ifstream f(FAVORITES_FILE);
    if (f.is_open()) {
        json j;
        f >> j;
        if (j.contains("pairs")) {
            for (auto& p : j["pairs"]) {
                pairs.push_back({p["base"], p["target"]});
            }
        }
        f.close();
    }
    if (pairs.empty()) {
        pairs = {{"USD", "EUR"}, {"USD", "JPY"}};
    }
    return pairs;
}

void saveFavorites(const vector<Pair>& pairs) {
    json j;
    j["pairs"] = json::array();
    for (auto& p : pairs) {
        j["pairs"].push_back({{"base", p.base}, {"target", p.target}});
    }
    ofstream f(FAVORITES_FILE);
    f << setw(2) << j << endl;
}

map<string, double> getRates(const string& base) {
    string url = API_URL + base;
    string resp = fetchUrl(url);
    if (resp.empty()) return {};
    auto data = json::parse(resp);
    if (!data.contains("rates")) return {};
    map<string, double> rates;
    for (auto& [key, val] : data["rates"].items()) {
        rates[key] = val.get<double>();
    }
    return rates;
}

void listRates(const vector<Pair>& pairs) {
    if (pairs.empty()) {
        cout << "No favorite pairs.\n";
        return;
    }
    // group by base
    map<string, vector<Pair>> grouped;
    for (auto& p : pairs) {
        grouped[p.base].push_back(p);
    }
    cout << "\n💱 Favorite Currency Pairs\n\n";
    for (auto& [base, basePairs] : grouped) {
        auto rates = getRates(base);
        for (auto& p : basePairs) {
            if (rates.count(p.target)) {
                cout << base << "/" << p.target << ": " << fixed << setprecision(4) << rates[p.target] << "\n";
            } else {
                cout << base << "/" << p.target << ": N/A\n";
            }
        }
    }
    time_t t = time(nullptr);
    cout << "\nUpdated: " << put_time(localtime(&t), "%Y-%m-%d %H:%M:%S") << "\n";
}

int main(int argc, char* argv[]) {
    curl_global_init(CURL_GLOBAL_DEFAULT);
    if (argc < 2) {
        auto pairs = loadFavorites();
        listRates(pairs);
        curl_global_cleanup();
        return 0;
    }
    string cmd = argv[1];
    if (cmd == "list") {
        auto pairs = loadFavorites();
        listRates(pairs);
    } else if (cmd == "add") {
        if (argc != 4) {
            cerr << "Usage: add BASE TARGET\n";
            curl_global_cleanup();
            return 1;
        }
        string base = argv[2];
        string target = argv[3];
        transform(base.begin(), base.end(), base.begin(), ::toupper);
        transform(target.begin(), target.end(), target.begin(), ::toupper);
        auto pairs = loadFavorites();
        bool exists = false;
        for (auto& p : pairs) {
            if (p.base == base && p.target == target) {
                exists = true;
                break;
            }
        }
        if (!exists) {
            pairs.push_back({base, target});
            saveFavorites(pairs);
            cout << "✅ Added " << base << "/" << target << "\n";
        } else {
            cout << "Pair " << base << "/" << target << " already in favorites.\n";
        }
    } else if (cmd == "remove") {
        if (argc != 4) {
            cerr << "Usage: remove BASE TARGET\n";
            curl_global_cleanup();
            return 1;
        }
        string base = argv[2];
        string target = argv[3];
        transform(base.begin(), base.end(), base.begin(), ::toupper);
        transform(target.begin(), target.end(), target.begin(), ::toupper);
        auto pairs = loadFavorites();
        vector<Pair> newPairs;
        bool removed = false;
        for (auto& p : pairs) {
            if (p.base == base && p.target == target) {
                removed = true;
            } else {
                newPairs.push_back(p);
            }
        }
        if (removed) {
            saveFavorites(newPairs);
            cout << "✅ Removed " << base << "/" << target << "\n";
        } else {
            cout << "Pair " << base << "/" << target << " not found.\n";
        }
    } else if (cmd == "watch") {
        int interval = 60;
        for (int i = 2; i < argc; i++) {
            if (string(argv[i]) == "--interval" && i+1 < argc) {
                interval = stoi(argv[i+1]);
                i++;
            }
        }
        cout << "Watching every " << interval << "s. Press Ctrl+C to stop.\n";
        auto pairs = loadFavorites();
        while (true) {
            listRates(pairs);
            this_thread::sleep_for(chrono::seconds(interval));
        }
    } else {
        cerr << "Usage: currency_favorites [list|add BASE TARGET|remove BASE TARGET|watch [--interval N]]\n";
        curl_global_cleanup();
        return 1;
    }
    curl_global_cleanup();
    return 0;
}
